### PR TITLE
[Feature] Add thrs in eval_metrics

### PR DESCRIPTION
**Pull request by LXXXXR**
Hi,
This pull request add a argument thr in accuracy, precision, recall and f1 score to support evaluation under certain threshold conditions. 
Please take a look. Thank you.

### CODE REVIEW

File: mmcls/core/evaluation/eval_metrics.py [hunk 1]
```diff
@@ -32,105 +34,144 @@ def calculate_confusion_matrix(pred, target):
...
+            or a np.array with shape C if the average is set to none.
+    """
+
+    allowed_average = ['macro', 'none']
+    if average not in allowed_average:
+        raise ValueError(f'Unsupport type of averaging {average}.')
+
+    if isinstance(pred, torch.Tensor) and isinstance(target, torch.Tensor):
+        pred = pred.detach().cpu().numpy()
+        target = target.detach().cpu().numpy()
```
Comment (MeowZheng) [inline mmcls/core/evaluation/eval_metrics.py:62]: pred and target are not cuda or leaf tensors by default.

**Review by MeowZheng — APPROVED** [review]

File: mmcls/core/evaluation/eval_metrics.py [hunk 2]
```diff
@@ -32,105 +34,144 @@ def calculate_confusion_matrix(pred, target):
...
+
+    allowed_average = ['macro', 'none']
+    if average not in allowed_average:
+        raise ValueError(f'Unsupport type of averaging {average}.')
+
+    if isinstance(pred, torch.Tensor) and isinstance(target, torch.Tensor):
+        pred = pred.numpy()
+        target = target.numpy()
+    elif not (isinstance(pred, np.ndarray) and isinstance(target, np.ndarray)):
+        raise TypeError('pred and target should both be'
```
Comment (ZwwWayne) [inline mmcls/core/evaluation/eval_metrics.py:64]: What if pred is ndarray and target is tensor?

File: mmcls/core/evaluation/eval_metrics.py [hunk 3]
```diff
@@ -32,105 +34,144 @@ def calculate_confusion_matrix(pred, target):
...
+    gt_positive = label == target.reshape(-1, 1)
+    precision = (pred_positive & gt_positive).sum(0) / np.maximum(
+        pred_positive.sum(0), 1) * 100
+    recall = (pred_positive & gt_positive).sum(0) / np.maximum(
+        gt_positive.sum(0), 1) * 100
+    f1_score = 2 * precision * recall / np.maximum(precision + recall, 1e-20)
+    if average == 'macro':
+        precision = float(precision.mean())
+        recall = float(recall.mean())
+        f1_score = float(f1_score.mean())
```
Comment (ZwwWayne) [inline mmcls/core/evaluation/eval_metrics.py:85]: else:?

**Review by ZwwWayne — APPROVED** [review]

File: mmcls/core/evaluation/eval_metrics.py [hunk 4]
```diff
@@ -6,22 +6,25 @@ def calculate_confusion_matrix(pred, target):
...
+    if isinstance(target, np.ndarray):
         target = torch.from_numpy(target)
-    elif not (isinstance(pred, torch.Tensor)
-              and isinstance(target, torch.Tensor)):
-        raise TypeError('pred and target should both be'
-                        'torch.Tensor or np.ndarray')
-    _, pred_label = pred.topk(1, dim=1)
+    assert (isinstance(pred, torch.Tensor)
+            and isinstance(target, torch.Tensor)), (
+                'pred and target should be torch.Tensor or np.ndarray')
```
Comment (ZwwWayne) [inline mmcls/core/evaluation/eval_metrics.py:24]: show type of pred and target in error message.

File: mmcls/core/evaluation/eval_metrics.py [hunk 5]
```diff
@@ -32,105 +35,146 @@ def calculate_confusion_matrix(pred, target):
...
+        pred_positive.sum(0), 1) * 100
+    recall = (pred_positive & gt_positive).sum(0) / np.maximum(
+        gt_positive.sum(0), 1) * 100
+    f1_score = 2 * precision * recall / np.maximum(precision + recall, 1e-20)
+    if average == 'macro':
+        precision = float(precision.mean())
+        recall = float(recall.mean())
+        f1_score = float(f1_score.mean())
+    else:
+        pass
```
Comment (ZwwWayne) [inline mmcls/core/evaluation/eval_metrics.py:88]: ？Should not we raise a warning or error?

File: mmcls/core/evaluation/eval_metrics.py [hunk 6]
```diff
@@ -32,105 +35,146 @@ def calculate_confusion_matrix(pred, target):
...
-def precision(pred, target, average='macro'):
+def precision_recall_f1(pred, target, average='macro', thr=None):
+    """Calculate precision, recall and f1 score according to the prediction and
+         target.
+
+    Args:
+        pred (torch.Tensor | np.array): The model prediction with shape (N, C).
+        target (torch.Tensor | np.array): The target of each prediction with
+            shape (N, 1) or (N,).
+        average (str): The type of averaging performed on the result.
```
Comment (ZwwWayne) [inline mmcls/core/evaluation/eval_metrics.py:46]: average -> average_mode

File: mmcls/core/evaluation/eval_metrics.py [hunk 7]
```diff
@@ -32,105 +35,146 @@ def calculate_confusion_matrix(pred, target):
...
+def precision(pred, target, average='macro', thr=None):
     """Calculate precision according to the prediction and target.
 
     Args:
-        pred (torch.Tensor | np.array): The model prediction.
-        target (torch.Tensor | np.array): The target of each prediction.
+        pred (torch.Tensor | np.array): The model prediction with shape (N, C).
+        target (torch.Tensor | np.array): The target of each prediction with
+            shape (N, 1) or (N,).
         average (str): The type of averaging performed on the result.
```
Comment (ZwwWayne) [inline mmcls/core/evaluation/eval_metrics.py:100]: See comments above.

**Comment by LXXXXR** [comment]
> Hi,
> This pull request add a argument thr in accuracy, precision, recall and f1 score to support evaluation under certain threshold conditions.
> Please take a look. Thank you.

I have modified thr to thrs according to comments. 
Pls see the unittest for what the evaluation results would look like. Thank you.
My concern is that the key is a little bit complicated, so pls let me know if you find it confusing. The reason why the functions that calculate the metrics can not return a dict is that the return format for accuracy is fixed and used in heads. I feel that changing that might cause troubles to users and it is best if other metrics return the same format as accuracy.
It has come to my attention that we also need to modify test.py. And it might need further modifications when we support multilabel feature. Therefore I will later submit another pull request to fix that after this pull request and #145 are merged.

**Review by ZwwWayne — APPROVED** [review]

**Review by ycxioooong — APPROVED** [review]
LGTM on multi-threshold part
